module/zfs/zio.c

```diff
@@ -3787,6 +3787,40 @@ zio_alloc_zil(spa_t *spa, objset_t *os, uint64_t txg, blkptr_t *new_bp,
  * ==========================================================================
  */
 
+static void
+zio_spa_config_enter(zio_t *zio)
+{
+	spa_t *spa = zio->io_spa;
+
+	metaslab_class_t *mc = spa_normal_class(spa);
+	if (mc->mc_ops->msop_block_based || zio->io_type != ZIO_TYPE_WRITE) {
+		spa_config_enter(spa, SCL_ZIO, zio, RW_READER);
+		return;
+	}
+
+	/*
+	 * Object based pools may need to continue to push I/Os even
+	 * if there is writer waiting for the SCL_ZIO lock. When a
+	 * writer starts to wait for the spa_config_lock, we will
+	 * set the maximum offset that we will allow to bypass
+	 * the starvation logical. Any I/Os that less than or
+	 * equal to that offset will be allowed to grab the reader
+	 * lock even though a writer is waiting.
```

Comment:
I'm having trouble parsing this, especially "starvation logical", is this correct?

```suggestion
	 * Object based pools may need to continue to push I/Os even
	 * if there is writer waiting for the SCL_ZIO lock because ....?. When a
	 * writer waits for the spa_config_lock, it will remember the last allocated BlockId.
	 * These already-allocated blocks will be allowed to bypass the normal writer priority logic
	 * and acquire the reader
	 * lock even though a writer is waiting.  This limited reader priority will not starve writers
	 * because it only applies to zio's that are already in progress (have already allocated their BlockId).
	 * New zio's will have to wait for the writer to acquire the lock before they can acquire a read lock.
```

Reply:
Yes, thanks for clarifying.